src/Microsoft.Identity.Client/Telemetry.cs

```diff
@@ -42,11 +45,144 @@ public void RegisterReceiver(Receiver r)
 
         private static readonly Telemetry Singleton = new Telemetry();
 
-        private Telemetry(){}
+        internal Telemetry(){}  // This is an internal constructor to build isolated unit test instance
 
         public static Telemetry GetInstance()
         {
             return Singleton;
         }
+
+        public bool TelemetryOnFailureOnly { get; set; }
+
+        private ConcurrentDictionary<Tuple<string, string>, EventBase> EventsInProgress = new ConcurrentDictionary<Tuple<string, string>, EventBase>();
+
+        private ConcurrentDictionary<string, List<EventBase>> CompletedEvents = new ConcurrentDictionary<string, List<EventBase>>();
+
+        internal string GenerateNewRequestId()
+        {
+            return Guid.NewGuid().ToString();
+        }
+
+        internal void StartEvent(string requestId, EventBase eventToStart)
+        {
+            if (_receiver != null && requestId != null)
+            {
+                EventsInProgress[new Tuple<string, string>(requestId, eventToStart[EventBase.ConstEventName])] = eventToStart;
+            }
+        }
+
+        internal void StopEvent(string requestId, EventBase eventToStop)
+        {
+            if (_receiver == null || requestId == null)
+            {
+                return;
+            }
+            Tuple<string, string> eventKey = new Tuple<string, string>(requestId, eventToStop[EventBase.ConstEventName]);
+
+            // Locate the same name event in the EventsInProgress map
+            EventBase eventStarted = null;
+            if (EventsInProgress.ContainsKey(eventKey))
+            {
+                eventStarted = EventsInProgress[eventKey];
+            }
+
+            // If we did not get anything back from the dictionary, most likely its a bug that StopEvent
+            // was called without a corresponding StartEvent
+            if (null == eventStarted)
+            {
+                // Stop Event called without a corresponding start_event.
+                return;
+            }
+
+            // Set execution time properties on the event
+            eventToStop.Stop();
+
+            if (!CompletedEvents.ContainsKey(requestId))
+            {
+                // if this is the first event associated to this
+                // RequestId we need to initialize a new List to hold
+                // all of sibling events
+                List<EventBase> events = new List<EventBase>();
+                events.Add(eventToStop);
+                CompletedEvents[requestId] = events;
+            }
+            else
+            {
+                // if this event shares a RequestId with other events
+                // just add it to the List
+                CompletedEvents[requestId].Add(eventToStop);
+            }
+
+            // Mark this event as no longer in progress
+            EventBase dummy;
+            EventsInProgress.TryRemove(eventKey, out dummy);
```

Comment:
That one was the alternative answer mentioned in the [StackOverflow post](http://stackoverflow.com/questions/7464965/why-does-concurrentdictionary-tryremove-require-a-second-out-argument) which I quoted in the [previous conversation](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/314/files#r111843592). It is fair to say this topic is tricky (too tricky than it should be), so let's take a step back and see the big picture here.

First of all, `ConcurrentDictionary dict` does not provide a straightforward `dict.Remove(key)` to do this (which I still don't understand the design decision behind that), now basically there are 2 options:

    // Use a standard method TryRemove()
    EventBase dummy;
    EventsInProgress.TryRemove(eventKey, out dummy);

    // Call a shadowed method, by first casting your instance into another object. Otherwise your code won't compile.
    ((IDictionary)dict).Remove("mykey");

The `dummy` variable in the first solution looks ugly (I hate that too), but at least reviewers would know its intention, and know the implementation logic is correct.

The casting behavior in the second one, I assume, would cause most reviewer to feel nervous, and start to check the ConcurrentDictionary's doc to see what is its type, what it inherits, what interface(s) it implements. All those are arguably the implementation details that a caller shouldn't have need to care. That's the downside I would try to avoid. So the current implementation chose the first approach.

As a side note, @kpanwar gave a suggestion on assigning a `null` to the `dummy` before using it. I'll include it in next commit.